compiler/circle-opselector/src/OpSelector.cpp

```diff
@@ -0,0 +1,396 @@
+{
+  auto graph = gb_context.graph();
+  auto nodefinder = gb_context.nodefinder();
+
+  const auto &tensors = _reader.tensors();
+  auto tensors_ptr = _reader.tensors_ptr();
+  assert(tensors_ptr != nullptr);
+
+  // graph outputs
+  const circle::TensorT &tensor = *tensors[output];
+
+  auto output_node = graph->nodes()->create<luci::CircleOutput>();
+  assert(output_node != nullptr);
+  auto output_from = nodefinder->node(output);
+  if (output_from != nullptr)
+    output_node->from(output_from);
+  else
+  {
+    // NOTE loco::Graph requires all input node(s) to a node should exist.
+    //      Here, CircleOutput needs an input node.
+    //      We add a dummy node to make it happy.
+    auto output_dummy = graph->nodes()->create<luci::CircleOutputDummy>();
+    assert(output_dummy != nullptr);
+    output_node->from(output_dummy);
+
+    luci::copy_tensor_attributes(tensor, output_dummy);
+    if (tensors_ptr->Get(output)->shape() == nullptr)
+      output_dummy->shape_status(luci::ShapeStatus::NOSHAPE);
+    else
+      output_dummy->shape_status(luci::ShapeStatus::VALID);
+  }
+
+  // set the graph output name and node object
+  auto graph_output = graph->outputs()->create();
+  std::string tname = luci::tensor_name(tensor);
+  assert(tname.length() > 0);
+  graph_output->name(tname);
+
+  luci::copy_tensor_attributes(tensor, output_node);
+
+  // Set GraphInputOutputIndex for graph
+  output_node->index(graph_output->index());
+
+  assert(tensor.shape_signature.size() == 0 ||
+         tensor.shape_signature.size() == tensor.shape.size());
+
+  // Shape of Output
+  auto output_shape = std::make_unique<loco::TensorShape>();
+  const std::vector<int32_t> &output_dims = tensor.shape; // in NHWC
+  output_shape->rank(output_dims.size());
+  for (uint32_t r = 0; r < output_dims.size(); ++r)
+  {
+    if (tensor.shape_signature.size() > 0 && tensor.shape_signature.at(r) == -1)
+      output_shape->dim(r).unset();
+    else
+      output_shape->dim(r).set(output_dims[r]);
+  }
+  graph_output->shape(std::move(output_shape));
+
+  // Data type
+  auto dtype = luci::luci_datatype(tensor.type);
+  graph_output->dtype(dtype);
+}
+
+std::unique_ptr<luci::Module>
+OpSelector::select_nodes(std::vector<const luci::CircleNode *> selected_nodes)
+{
+  auto module = luci::make_module();
+
+  const luci::GraphBuilderSource *source_ptr = &luci::GraphBuilderRegistry::get();
+
+  for (uint32_t g = 0; g < _reader.num_subgraph(); ++g)
+  {
+    std::unique_ptr<loco::Graph> graph = loco::make_graph(); // create new empty graph
+
+    assert(_reader.select_subgraph(g)); // select subgraph. usually, 0 is main
+    graph->name(_reader.name());        // set name of graph (if it has one graph, name is empty.)
+
+    auto nodefinder = std::make_unique<luci::IndexNodeFinder>();
+    auto tensoroutputs = std::make_unique<luci::IndexTensorOutputs>();
+
+    luci::GraphBuilderContext gb_context(graph.get(), &_reader, nodefinder.get(),
```

Comment:
Can you please explain why you need to use `GraphBuilderContext` ?

There's `Importer::importModule` what will read the circle file and return `Module`

Reply:
> Can you please explain why you need to use GraphBuilderContext ?

I use this because I create a new graph with only selected nodes to create a 'circle' file with selected nodes.

At first, I thought of two ways. The first was to use only module to perform Propagation of profiling data from the original module to the new module, and the second was to create a new graph by selecting only the desired data directly from the model data.
I tried it for quite a long time as the first method, but as you said, it was so difficult to understand the whole logic. So I chose the second way that was a little easier.

After the development was over, @mhs4670go informed me of the related module, so I was changing the structure, but the deadline was just around the corner, so I uploaded the existing code.